build-system/tasks/build-bento.js

```diff
@@ -0,0 +1,225 @@
+const argv = require('minimist')(process.argv.slice(2));
+const debounce = require('../common/debounce');
+const {
+  INABOX_EXTENSION_SET,
+  buildBinaries,
+  buildExtensionCss,
+  buildExtensionJs,
+  buildNpmBinaries,
+  buildNpmCss,
+  declareExtension,
+  dedupe,
+  getBentoBuildFilename,
+  getExtensionsFromArg,
+} = require('./extension-helpers');
+const {bentoBundles, verifyBentoBundles} = require('../compile/bundles.config');
+const {compileJison} = require('./compile-jison');
+const {endBuildStep, watchDebounceDelay} = require('./helpers');
+const {existsSync, mkdirSync} = require('fs');
+const {getBentoName} = require('./bento-helpers');
+const {log} = require('../common/logging');
+const {red} = require('kleur/colors');
+const {watch} = require('chokidar');
+
+// All declared components.
+const COMPONENTS = {};
+
+/**
+ * Initializes all components from build-system/compile/bundles.config.bento.json
+ * if not already done and populates the given components object.
+ * @param {?Object} componentsObject
+ * @param {boolean=} includeLatest
+ */
+function maybeInitializeComponents(componentsObject, includeLatest = false) {
+  if (Object.keys(componentsObject).length === 0) {
+    verifyBentoBundles();
+    bentoBundles.forEach((c) => {
+      declareExtension(
+        c.name,
+        c.version,
+        // TODO(rileyajones): Remove this once the bento build process fully seperated.
+        '0.1',
+        c.options,
+        componentsObject,
+        includeLatest
+      );
+    });
+  }
+}
+
+/**
+ * Process the command line arguments --nocomponents, --components, and
+ * --components_from and return a list of the referenced components.
+ *
+ * @param {boolean=} preBuild
+ * @return {!Array<string>}
+ */
+function getComponentsToBuild(preBuild = false) {
+  let componentsToBuild = [];
+  if (argv.extensions) {
+    if (typeof argv.extensions !== 'string') {
+      log(red('ERROR:'), 'Missing list of components.');
+      process.exit(1);
+    } else if (argv.extensions === 'inabox') {
+      argv.extensions = INABOX_EXTENSION_SET.join(',');
+    }
+    const explicitComponents = argv.extensions.replace(/\s/g, '').split(',');
+    componentsToBuild = dedupe(componentsToBuild.concat(explicitComponents));
```

Comment:
Almost every usage I've found of `dedupe` follows this pattern: `dedupe(arr1.concat(arr2))` or `dedupe([...arr1, ...arr2])`
It's currently implemented as
```

/**
 * Remove duplicates from the given array.
 * @param {!Array<string>} arr
 * @return {!Array<string>}
 */
function dedupe(arr) {
  const map = Object.create(null);
  arr.forEach((item) => (map[item] = true));
  return Object.keys(map);
}
```

It could be
```
/**
 * Remove duplicates from the given array.
 * @param {!Array<string>...} arrs
 * @return {!Array<string>}
 */
function dedupe(arrs...) {
  const s = new Set();
  arrs.forEach(arr => s.add(...arr));
  return s.values();
}
```

Still doing pretty much the same thing, but would make for simple call-sites

Reply:
As a totally separate note, have you considered making `componentsToBuild` itself a `Set`? We repeatedly concat+dedupe here (albeit in mutually exclusive `if`s IIUC), a set might be more semantic

Reply:
As to why this is written this way, I didn't really write this so I can't really say. However, using a set definitely makes sense here.